source/nodejs/adaptivecards/src/card-elements.ts

```diff
@@ -2953,6 +2953,208 @@ export class TextInput extends Input {
     }
 }
 
+export class FileResult extends SerializableObject {
+    //#region Schema
+
+    static readonly fileNameProperty = new StringProperty(Versions.v1_0, "fileName");
+    static readonly contentProperty  = new StringProperty(Versions.v1_0, "content");
+
+    @property(FileResult.fileNameProperty)
+    fileName?: string;
+
+    @property(FileResult.contentProperty)
+    content?: string;
+
+    //#endregion
+
+    protected getSchemaKey(): string {
+        return "FileResult";
+    }
+
+    constructor(fileName?: string, content?: string) {
+        super();
+
+        this.fileName = fileName;
+        this.content = content;
+    }
+}
+
+export class FileInput extends Input {
+  //#region Schema
+
+  static readonly valueProperty = new SerializableObjectProperty(Versions.v1_0, "value", FileResult);
+  static readonly maxLengthProperty = new NumProperty(Versions.v1_0, "maxLength");
+  static readonly placeholderProperty = new StringProperty(Versions.v1_0, "placeholder");
+  static readonly styleProperty = new EnumProperty(Versions.v1_0, "style", Enums.InputTextStyle, Enums.InputTextStyle.File);
+  static readonly regexProperty = new StringProperty(Versions.v1_3, "regex", true);
+
+  @property(FileInput.valueProperty)
+  defaultValue?: FileResult;
+
+  @property(FileInput.maxLengthProperty)
+  maxLength?: number;
+
+  @property(FileInput.placeholderProperty)
+  placeholder?: string;
+
+  @property(FileInput.styleProperty)
+  style: Enums.InputTextStyle = Enums.InputTextStyle.File;
+
+  @property(FileInput.regexProperty)
+  regex?: string;
+
+  //#endregion
+
+  private setupInput(input: HTMLInputElement) {
+      input.style.flex = "1 1 auto";
+      input.tabIndex = 0;
+
+      if (this.placeholder) {
+          input.placeholder = this.placeholder;
+          input.setAttribute("aria-label", this.placeholder)
+      }
+
+      if (this.maxLength && this.maxLength > 0) {
+          input.maxLength = this.maxLength;
+      }
+
+      input.oninput = () => {
+          this.valueChanged();
+          let fileName = (<HTMLInputElement>this.renderedInputControlElement).value;
+          let file = (<HTMLInputElement>this.renderedInputControlElement).files?.item(0) as File;
+          let fileReader = new FileReader();
+          fileReader.onload = () => {
+              let content = fileReader.result as string;
+              this.defaultValue = new FileResult(fileName, content);
```

Comment:
For me this is a big problem with this implementation - the file is entirely read, converted to base64 and stored in memory. Then it is passed as a bsed64 encoded string to the backend when an Action.Submit or Action.Execute is executed. I don't think that scales beyond very small files.

IMO the right model here is more complicated. We need to have a way to let the host application get access to the byte stream of the selected file at the time the action is executed to allow it to handle the transfer of the bytes in an optimal way, which will most likely involve in many cases sending the file in chunks in order to never have to read it all into memory. The problem obviously is that making all of this work requires explicit client support, which currently doesn't exist.

This implementation definitely has the benefit of working with WebChat/Bots "as is" but I don't believe it is the right solution.

Reply:
Totally agree with you, i had no choice there for a first version of that feature, i saw in the upload process of botframework-emulator that files are stored as attachments. I don't really have all the spec on that part and decided to make a simple start. Still have nothing more to tell on how to change that part for now.